### PR TITLE
PRIME-2359 Enhance Site Update Event Log

### PR DESCRIPTION
**Pull request by bergomi02**
https://pr-2235.pharmanetenrolment.gov.bc.ca/site

### CODE REVIEW

**Comment by neophyte57** [comment]
(1) Before I even submitted the Site for Review, the Event Log was filling up with "Site Updated" events, not desired behavior?

image

(2) There is something strange with this output 
- CareConnect is removed.
- CareConnect is added.

(3) There is something strange with this output 
- Physical Address changed from 
      BC CA 
  to
   1-2600 Beaverbrook Cres  Burnaby BC CA V3J 7W6.

**Comment by bergomi02** [comment]
> (1) Before I even submitted the Site for Review, the Event Log was filling up with "Site Updated" events, not desired behavior?
> 
> image
> 
> (2) There is something strange with this output
> 
> * CareConnect is removed.
> * CareConnect is added.
> 
> (3) There is something strange with this output
> 
> * Physical Address changed from
>   BC CA
>   to
>   1-2600 Beaverbrook Cres  Burnaby BC CA V3J 7W6.

This should be fixed.

**Comment by neophyte57** [comment]
I made changes to "SMS Phone", which were saved, but the change was not noted in Event Log 

image

File: prime-dotnet-webapi/Services/CommunitySiteService.cs [hunk 1]
```diff
@@ -310,25 +424,30 @@ private bool UpdateRemoteUsers(Site current, IEnumerable<SiteRemoteUserUpdateMod
                     newRemoteUser.SiteId = current.Id;
                     _context.RemoteUsers.Add(newRemoteUser);
 
-                    remoteUsersUpdated ^= true;
+                    result.Add($"Remote user '{updatedUser.FirstName} {updatedUser.LastName}' is added.");
```
[the remark this answers is not in the document]
Comment (bergomi02) [reply prime-dotnet-webapi/Services/CommunitySiteService.cs:427]: The wording is updated.

**Comment by neophyte57** [comment]
Somewhat out of scope but still better to include in this PR:  If Admin changes Site ID, there is a Event Log message of "Site ID (PEC Code) associated with site" but it should identify previous value and new Site ID.  Can leave out "PEC Code" wording as Leanne said it is outdated terminology.

File: prime-dotnet-webapi/Services/CommunitySiteService.cs [hunk 2]
```diff
@@ -244,38 +281,112 @@ private void UpdateContacts(CommunitySite current, CommunitySiteUpdateModel upda
...
+            var result = new List<string>();
+            var propertyNames = new[,]
+            {
+                {nameof(currentContact.FirstName), "First Name"},
+                {nameof(currentContact.LastName), "Last Name"},
+                {nameof(currentContact.JobRoleTitle), "Job Title"},
+                {nameof(currentContact.Email), "Email"},
+                {nameof(currentContact.Phone), "Phone"},
+                {nameof(currentContact.Fax), "Fax"},
+                {nameof(currentContact.SMSPhone), "SMS Phone"},
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/CommunitySiteService.cs:309]: For some reason, my change to SMS Phone was not captured in the Event Log

File: prime-dotnet-webapi/Services/CommunitySiteService.cs [hunk 3]
```diff
@@ -244,38 +281,112 @@ private void UpdateContacts(CommunitySite current, CommunitySiteUpdateModel upda
...
             }
+            return result;
         }
 
-        private void UpdateBusinessHours(Site current, CommunitySiteUpdateModel updated)
+        private static string TranslateContactType(string contactType)
         {
+            return contactType switch
+            {
+                "AdministratorPharmaNet" => "Administrator PharmaNet",
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/CommunitySiteService.cs:291]: Should be "PharmaNet Administrator"

**Comment by bergomi02** [comment]
> Somewhat out of scope but still better to include in this PR: If Admin changes Site ID, there is a Event Log message of "Site ID (PEC Code) associated with site" but it should identify previous value and new Site ID. Can leave out "PEC Code" wording as Leanne said it is outdated terminology.

The message has been updated to similar language to the vendor update message.

File: prime-dotnet-webapi/Services/CommunitySiteService.cs (hunk 2 above)
Comment (bergomi02) [reply prime-dotnet-webapi/Services/CommunitySiteService.cs:309]: It is fixed.

File: prime-dotnet-webapi/Services/CommunitySiteService.cs (hunk 3 above)
Comment (bergomi02) [reply prime-dotnet-webapi/Services/CommunitySiteService.cs]: It is fixed.

**Comment by neophyte57** [comment]
I made no explicit change to Hours of Operation but seeing a message 
image

**Comment by bergomi02** [comment]
> I made no explicit change to Hours of Operation but seeing a message image

It is fixed

**Comment by bergomi02** [comment]
> (a) 24-Hours Operation looks weird, (b) also formatting for times should include hyphen, e.g. "Tuesday 09:00:00-17:00:00"
> 
> image

It is fixed. Also, Remote User email update is implemented.

**Comment by bergomi02** [comment]
> Re: image
> 
> 1. The format of the hours should be consistent with how they are displayed in the overview, i.e. no seconds portion
>    image
> 2. For added hours (Wednesday and Friday in my case), Event Log should indicate they were added

1) It is fixed. Also put more space in between segments to make it easier to read.
2) I disagree about adding text to indicate what is added (or updated). It shows before and after, and it is pretty easy to tell the difference. I think adding more text will make it clouded and not necessary. Maybe we can ask the client about it.

**Comment by neophyte57** [comment]
> > Re: image
> > 
> > 1. The format of the hours should be consistent with how they are displayed in the overview, i.e. no seconds portion
> >    image
> > 2. For added hours (Wednesday and Friday in my case), Event Log should indicate they were added
> 
> 1. It is fixed. Also put more space in between segments to make it easier to read.
> 2. I disagree about adding text to indicate what is added (or updated). It shows before and after, and it is pretty easy to tell the difference. I think adding more text will make it clouded and not necessary. Maybe we can ask the client about it.

The changes look much better ... never mind about #2 (I think I was just confused) 

image
versus 
image

File: prime-dotnet-webapi/Services/CommunitySiteService.cs [hunk 4]
```diff
@@ -12,6 +12,7 @@
 using Prime.ViewModels;
 using AutoMapper.QueryableExtensions;
 using Prime.Models.Api;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/CommunitySiteService.cs:15]: Unnecessary

Comment (bergomi02) [reply prime-dotnet-webapi/Services/CommunitySiteService.cs]: reference removed

File: prime-dotnet-webapi/Services/CommunitySiteService.cs [hunk 5]
```diff
@@ -151,28 +156,39 @@ public async Task UpdateSiteAsync(int siteId, CommunitySiteUpdateModel updatedSi
 
             _context.Entry(currentSite).CurrentValues.SetValues(updatedSite);
 
+            var updateDetail = new List<string>();
             if (currentSite.SubmittedDate == null)
             {
-                UpdateVendors(currentSite, updatedSite);
+                var vendors = await GetVendorsAsync();
+                UpdateVendors(currentSite, updatedSite, vendors);
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/CommunitySiteService.cs:163]: `UpdateVendors` returns change details but they are discarded?

Comment (bergomi02) [reply prime-dotnet-webapi/Services/CommunitySiteService.cs]: Good catch. It is updated.

Comment (neophyte57) [reply prime-dotnet-webapi/Services/CommunitySiteService.cs]: The change has a side-effect 
image

Comment (neophyte57) [reply prime-dotnet-webapi/Services/CommunitySiteService.cs]: @bergomi02 FYI I'm going to revert the change, add comment, and continue PR review

File: prime-dotnet-webapi/Services/CommunitySiteService.cs [hunk 6]
```diff
@@ -191,25 +209,40 @@ public async Task<PermissionsRecord> GetPermissionsRecordAsync(int siteId)
...
             if (updated.PhysicalAddress == null)
             {
-                return;
+                return result;
             }
 
             if (current.PhysicalAddress == null)
             {
                 current.PhysicalAddress = updated.PhysicalAddress;
+                result.Add($"Physcial Address '{AddressToString(updated.PhysicalAddress)}' was added.");
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/CommunitySiteService.cs:223]: Typo in `Physcial`

Comment (bergomi02) [reply prime-dotnet-webapi/Services/CommunitySiteService.cs]: It is fixed

**Review by neophyte57 — APPROVED** [review]
Nice job! 👍  A few small issues remain but shouldn't prevent going to PROD IMO.  I also made some minor commits you may want to review.

**Pull request merged**